tsdb/db.go

```diff
@@ -1505,34 +1505,38 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
-				dir := filepath.Join(db.Dir(), uid.String())
-				if err := os.RemoveAll(dir); err != nil {
-					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
-				}
-			}
-		}
-	}()
+	// We can momentarily allow overlapping blocks in this particular case.
+	// This will allow the reload function to replace old blocks with the new ones.
+	// Notice that the the final result will be overlapping blocks, but with no tombstones.
+	if !db.opts.AllowOverlappingBlocks {
+		db.opts.AllowOverlappingBlocks = true
+		defer func() { db.opts.AllowOverlappingBlocks = false }()
+	}
 
-	db.mtx.RLock()
-	blocks := db.blocks[:]
-	db.mtx.RUnlock()
+	cleanUpCompleted := false
+	// Repeat cleanup until there is no tombstones left.
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
 
-	for _, b := range blocks {
-		if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
-			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
-			return err
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
-		}
-	}
+		for _, pb := range db.Blocks() {
+			uid, cleanErr := pb.CleanTombstones(db.Dir(), db.compactor)
+			if cleanErr != nil {
+				return errors.Wrapf(cleanErr, "clean tombstones: %s", pb.Dir())
+			}
+			if uid == nil {
+				// There was nothing to clean
+				continue
+			}
 
-	if err := db.reloadBlocks(); err != nil {
-		return errors.Wrap(err, "reload blocks")
+			pb.meta.Compaction.Deletable = true
```

Comment:
We don't need this since the new block has the old block as a parent and hence will be deleted.

Reply:
That's what I thought too, but to my surprise, they are not deleted without this line

Reply:
I saw your comment in the email saying `That's what I thought too, but to my surprise, they are not deleted without this line`, and I think that is possible! If the tombstones covered the entire block, then there would be no resultant block to tell the parent. So having this here makes sense.

Reply:
You can add a comment describing why it is required (the reason that I stated above)

Reply:
Yeah, I've written that but then remembered that on test "TestTombstoneClean" it works just fine without it. I deleted the comment just because I wanted to investigate a little bit more. 

What you mention above is exactly what happens on "TestTombstoneCleanFail", the new block is created without a parent, and the old block is empty :+1: